clippy_lints/src/methods/iter_overeager_cloned.rs

```diff
@@ -0,0 +1,78 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet;
+use clippy_utils::ty::implements_trait;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_lint::LateContext;
+use rustc_span::sym;
+
+use super::ITER_OVEREAGER_CLONED;
+use crate::redundant_clone::REDUNDANT_CLONE;
+
+/// lint use of `cloned().last()` for `Iterators`
+pub(super) fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx hir::Expr<'_>,
+    recv: &'tcx hir::Expr<'_>,
+    name: &str,
+    map_arg: &[hir::Expr<'_>],
+) {
+    // lint if caller of `.cloned().last()` is an Iterator
+    let recv_impls_iterator = cx.tcx.get_diagnostic_item(sym::Iterator).map_or(false, |id| {
+        implements_trait(cx, cx.typeck_results().expr_ty(recv), id, &[])
+    });
+    if recv_impls_iterator {
+        let (lint, msg) = match name {
+            "count" => (
+                REDUNDANT_CLONE,
+                format!(
+                    "called `cloned().{}()` on an `Iterator`. It may be more efficient to call\
+                    `.{}()` instead",
+                    name, name
```

Comment:
@llogiq I'm not sure if this is the preferred way to return `REDUNDANT_CLONE` lint.

If not I'll change the code of `redundant_clone.rs`.

Reply:
Why would it not be? Issuing the lint here keeps the code simple. Perhaps add a comment to the `redundant_clone` code that links to the implementation.

Reply:
> Why would it not be? Issuing the lint here keeps the code simple. Perhaps add a comment to the `redundant_clone` code that links to the implementation.

@llogiq This depends on `clippy` design; Whenever or or all `check` always run or not. I suspect that's the case, but I didn't read that part of the code to make sure.

Reply:
And also on how checks are used. I don't know whenever, there is a rule that says "all lints should be generated from file with the same name". It was the case for some lints, but I didn't check if that's true. Though, I see you don't see a problem, so I guess it's fine.